### PR TITLE
Add calico to Kubernetes addon

### LINKED ISSUE #436: [ADDON] - Calico
Add support for Calico as an add-on.

https://projectcalico.docs.tigera.io/getting-started/kubernetes/helm

### PR DESCRIPTION
**Pull request by florentio**
### What does this PR do?

Adds the [Calico](https://projectcalico.docs.tigera.io/getting-started/kubernetes/helm) Helm chart to the Kubernetes addons. 

### Motivation

My company are in the process of adopting calico for networking and network security solution for our EKS cluster. This addon would be a useful addon for us to deploy to our clusters. 
Also there is already an issue ticket on the roadmap related to it https://github.com/aws-ia/terraform-aws-eks-blueprints/issues/436

- Closes #436

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [x] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators

- [ ] E2E Test successfully complete before merge?

### Additional Notes

### CODE REVIEW

File: docs/add-ons/calico.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# CALICO
```
Comment (bryantbiggs) [inline docs/add-ons/calico.md:1]:
```suggestion
# Calico
```

Lets ensure these are all title cased

File: docs/add-ons/calico.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    namespace  = "tigera-operator"                                      # (Optional) The namespace to install the release into.
+    values = [templatefile("${path.module}/calico-values.yaml", {})]
+  }
+```
+
+### GitOps Configuration
+
+The following properties are made available for use when managing the add-on via GitOps.
+
+```
```
Comment (bryantbiggs) [inline docs/add-ons/calico.md:35]:
```suggestion
```sh
```

File: modules/kubernetes-addons/calico/locals.tf [hunk 3]
```diff
@@ -0,0 +1,27 @@
+locals {
+  name = "calico"
+  default_helm_config = {
+    name             = local.name
+    chart            = local.name
+    repository       = "https://projectcalico.docs.tigera.io/charts"
+    version          = "v3.24.0"
+    namespace        = "tigera-operator"
+    timeout          = "1200"
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/calico/locals.tf:9]: lets remove this, its the default for the helm addon module

File: modules/kubernetes-addons/calico/variables.tf [hunk 4]
```diff
@@ -0,0 +1,28 @@
+variable "helm_config" {
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/calico/variables.tf:1]: These should all be ordered
```hcl
variable "foo" {
  description = "First"
  type        = any
  default     = {}
}
```

**Review by bryantbiggs — COMMENTED** [review]
Looks great - few minor comments.

Once we merge https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/885 - lets update this PR to add the addon to that example to demonstrate and validate its usage

**Comment by florentio** [comment]
> Looks great - few minor comments.
> 
> Once we merge #885 - lets update this PR to add the addon to that example to demonstrate and validate its usage

Didn't you mean this one https://github.com/aws-ia/terraform-aws-eks-blueprints/pull/898 instead ?

**Comment by bryantbiggs** [comment]
> > Looks great - few minor comments.
> > Once we merge #885 - lets update this PR to add the addon to that example to demonstrate and validate its usage
> 
> Didn't you mean this one #898 instead ?

Ah yes, my apologies - correct!

File: modules/kubernetes-addons/calico/variables.tf [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+variable "manage_via_gitops" {
+  description = "Determines if the add-on should be managed via GitOps."
+  type        = bool
+  default     = false
+}
+
+variable "addon_context" {
+  type = object({
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/calico/variables.tf:14]: just make sure this one gets the ordering updated as well

File: docs/add-ons/calico.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+Deploy Calico with custom `values.yaml`
+
+```hcl
+  # Optional Map value; pass calico-values.yaml from consumer module
+   calico_helm_config = {
+    name       = "calico"                                               # (Required) Release name.
+    repository = "https://projectcalico.docs.tigera.io/charts"          # (Optional) Repository URL where to locate the requested chart.
+    chart      = "calico"                                               # (Required) Chart name to be installed.
```
Comment (bryantbiggs) [inline docs/add-ons/calico.md:24]:
```suggestion
    chart      = "tigera-operator"                                               # (Required) Chart name to be installed.
```

File: modules/kubernetes-addons/calico/locals.tf [hunk 7]
```diff
@@ -0,0 +1,26 @@
+locals {
+  name = "calico"
+  default_helm_config = {
+    name             = local.name
+    chart            = local.name
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/calico/locals.tf:5]:
```suggestion
    chart            = "tigera-operator"
```

File: modules/kubernetes-addons/calico/locals.tf [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    repository       = "https://projectcalico.docs.tigera.io/charts"
+    version          = "v3.24.0"
```
Comment (bryantbiggs) [inline modules/kubernetes-addons/calico/locals.tf:7]:
```suggestion
    version          = "v3.24.1"
```

**Review by bryantbiggs — COMMENTED** [review]
Few updates from testing. Per https://docs.aws.amazon.com/eks/latest/userguide/calico.html we should install `tigera-operator` and when I tried with the PR as it is currently, it fails to find the chart/install.

**Comment by florentio** [comment]
> Few updates from testing. Per https://docs.aws.amazon.com/eks/latest/userguide/calico.html we should install `tigera-operator` and when I tried with the PR as it is currently, it fails to find the chart/install.

Good catch. I am going to fix it. Thanks

**Review by bryantbiggs — APPROVED** [review]
Looks awesome, thank you!

**Pull request merged**